pkg/envoy/rds/response.go

```diff
@@ -56,5 +58,40 @@ func NewResponse(cataloger catalog.MeshCataloger, proxy *envoy.Proxy, _ *xds_dis
 		rdsResources = append(rdsResources, ingressRouteConfig)
 	}
 
+	if dr != nil {
+		// Ensure all RDS resources are responded to a given non-nil and non-empty request
+		// Empty RDS RouteConfig will be provided for resources requested that our logic did not fulfill
+		// due to policy logic
+		rdsResources = ensureRDSRequestCompletion(dr, rdsResources)
+	}
+
 	return rdsResources, nil
 }
+
+// ensureRDSRequestCompletion computes delta between requested resources and response resources.
+// If any resources requested were not responded to, this function will fill those in with empty RouteConfig stubs
+func ensureRDSRequestCompletion(dr *xds_discovery.DiscoveryRequest, rdsResources []types.Resource) []types.Resource {
+	requestMapset := mapset.NewSet()
+	for idx := range dr.ResourceNames {
+		requestMapset.Add(dr.ResourceNames[idx])
+	}
+
+	responseMapset := mapset.NewSet()
+	for _, res := range rdsResources {
+		responseMapset.Add(cache.GetResourceName(res))
+	}
+
+	// If there were any requested elements we didn't reply to, create empty/dummy RDS resources
+	// for those now
+	requestDifference := requestMapset.Difference(responseMapset)
+	requestDifference.Each(func(reqDif interface{}) bool {
+		unfulfilledRequestedResource := reqDif.(string)
+		rdsResources = append(rdsResources, route.NewRouteConfigurationStub(unfulfilledRequestedResource))
+
+		// Do not stop iterating until last elem
+		return false
+	})
```

Comment:
I love the functional approach `mapset` provides with `Each()`.  But I wonder if we can cut 3 lines by using `.Iterator()` - we won't have to `return false`.
```suggestion
	for reqDif := requestMapset.Difference(responseMapset).Iterator() {
		unfulfilledRequestedResource := reqDif.(string)
		rdsResources = append(rdsResources, route.NewRouteConfigurationStub(unfulfilledRequestedResource))
	}
```

Reply:
:ok_man: